In [1]:
import pandas as pd
import numpy as np
from ast import literal_eval

# Turn off warnings to ease reading
import warnings
warnings.filterwarnings('ignore')

# 0. Loading Datasets

In [2]:
vgc_data = pd.read_csv('Files/01_vgc_clean.csv', parse_dates = ['Release'])
rawg_data = pd.read_csv('Files/02_rawg_metacritic_url.csv', parse_dates = ['released'])
metacritic = pd.read_csv('Files/03_metacritic.csv')


### Converting lists interpreted as strings to lists ###

rawg_data['genres'] = rawg_data['genres'].apply(lambda x: literal_eval(x) if pd.notnull(x) and ('[' in x) else x)
rawg_data['stores'] = rawg_data['stores'].apply(lambda x: literal_eval(x) if pd.notnull(x) and ('[' in x) else x)
rawg_data['tags'] = rawg_data['tags'].apply(lambda x: literal_eval(x) if pd.notnull(x) and ('[' in x) else x)

In [3]:
vgc_data.head()

Title Platform         Publisher         Developer  \
0                      Minecraft      All            Mojang         Mojang AB   
1             Grand Theft Auto V      All    Rockstar Games    Rockstar North   
2                     Wii Sports      Wii          Nintendo      Nintendo EAD   
3  PlayerUnknown's Battlegrounds      All  PUBG Corporation  PUBG Corporation   
4              Super Mario Bros.      NES          Nintendo      Nintendo EAD   

     Release   Sales  
0 2011-11-18  200.00  
1 2013-09-17  140.00  
2 2006-11-19   82.90  
3 2017-12-17   70.00  
4 1985-10-18   40.24

In [4]:
rawg_data.head()

slug                name   released  suggestions_count  \
0  grand-theft-auto-v  Grand Theft Auto V 2013-09-17                416   
1  grand-theft-auto-v  Grand Theft Auto V 2013-09-17                416   
2  grand-theft-auto-v  Grand Theft Auto V 2013-09-17                416   
3  grand-theft-auto-v  Grand Theft Auto V 2013-09-17                416   
4  grand-theft-auto-v  Grand Theft Auto V 2013-09-17                416   

          platform               genres  \
0               PC  [Action, Adventure]   
1  Xbox Series S/X  [Action, Adventure]   
2    PlayStation 5  [Action, Adventure]   
3    PlayStation 4  [Action, Adventure]   
4    PlayStation 3  [Action, Adventure]   

                                              stores  \
0  [Epic Games, PlayStation Store, Xbox Store, Xb...   
1  [Epic Games, PlayStation Store, Xbox Store, Xb...   
2  [Epic Games, PlayStation Store, Xbox Store, Xb...   
3  [Epic Games, PlayStation Store, Xbox Store, Xb...   
4  [Epic Games, PlayStation Store, Xbox Store, Xb...   

                                                tags    esrb        plat_mc  \
0  [Singleplayer, Steam Achievements, Multiplayer...  Mature             pc   
1  [Singleplayer, Steam Achievements, Multiplayer...  Mature  xbox-series-x   
2  [Singleplayer, Steam Achievements, Multiplayer...  Mature  playstation-5   
3  [Singleplayer, Steam Achievements, Multiplayer...  Mature  playstation-4   
4  [Singleplayer, Steam Achievements, Multiplayer...  Mature            ps3   

                                                 url  
0  https://www.metacritic.com/game/pc/grand-theft...  
1  https://www.metacritic.com/game/xbox-series-x/...  
2  https://www.metacritic.com/game/playstation-5/...  
3  https://www.metacritic.com/game/playstation-4/...  
4  https://www.metacritic.com/game/ps3/grand-thef...

In [5]:
metacritic.head()

url  critic_score  \
0  https://www.metacritic.com/game/pc/grand-theft...          96.0   
1  https://www.metacritic.com/game/xbox-series-x/...           NaN   
2  https://www.metacritic.com/game/playstation-5/...           NaN   
3  https://www.metacritic.com/game/playstation-4/...          97.0   
4  https://www.metacritic.com/game/ps3/grand-thef...           NaN   

   critic_reviews  critic_pos  critic_mix  critic_neg user_score  \
0            57.0        56.0         1.0         0.0        7.8   
1             NaN         NaN         NaN         NaN        NaN   
2             NaN         NaN         NaN         NaN        NaN   
3            66.0        66.0         0.0         0.0        8.4   
4             NaN         NaN         NaN         NaN        NaN   

   user_reviews user_pos user_mix user_neg  
0        7286.0    5,401      831    1,054  
1           NaN      NaN      NaN      NaN  
2           NaN      NaN      NaN      NaN  
3        6414.0    5,282      554      578  
4           NaN      NaN      NaN      NaN

# 1. Removing Columns

We no longer need the next columns in `rawg_data`:

- `slug` : We only needed this column to create the url.

- `released` : We already have this column in the main `vgc_data` dataframe in the correct format (DD-MM-YYYY).

- `metacritic` : We have an up-to-date dataset with all this information.

- `plat_mc` : We only needed this column to create the url.

In [6]:
rawg = rawg_data.drop(['slug','released', 'plat_mc'], axis = 1)

In [7]:
rawg.head()

name  suggestions_count         platform  \
0  Grand Theft Auto V                416               PC   
1  Grand Theft Auto V                416  Xbox Series S/X   
2  Grand Theft Auto V                416    PlayStation 5   
3  Grand Theft Auto V                416    PlayStation 4   
4  Grand Theft Auto V                416    PlayStation 3   

                genres                                             stores  \
0  [Action, Adventure]  [Epic Games, PlayStation Store, Xbox Store, Xb...   
1  [Action, Adventure]  [Epic Games, PlayStation Store, Xbox Store, Xb...   
2  [Action, Adventure]  [Epic Games, PlayStation Store, Xbox Store, Xb...   
3  [Action, Adventure]  [Epic Games, PlayStation Store, Xbox Store, Xb...   
4  [Action, Adventure]  [Epic Games, PlayStation Store, Xbox Store, Xb...   

                                                tags    esrb  \
0  [Singleplayer, Steam Achievements, Multiplayer...  Mature   
1  [Singleplayer, Steam Achievements, Multiplayer...  Mature   
2  [Singleplayer, Steam Achievements, Multiplayer...  Mature   
3  [Singleplayer, Steam Achievements, Multiplayer...  Mature   
4  [Singleplayer, Steam Achievements, Multiplayer...  Mature   

                                                 url  
0  https://www.metacritic.com/game/pc/grand-theft...  
1  https://www.metacritic.com/game/xbox-series-x/...  
2  https://www.metacritic.com/game/playstation-5/...  
3  https://www.metacritic.com/game/playstation-4/...  
4  https://www.metacritic.com/game/ps3/grand-thef...

# 2. Merging RAWG and Metacritic Datasets

We will do this by joining both dataframes by the `pd.DataFrame.join()` method, using the `url` columns as indexes.

In [8]:
rawg = rawg.set_index('url')
metacritic = metacritic.set_index('url')

In [9]:
rawg_meta = rawg.join(metacritic)
rawg_meta.head()

name  \
url                                                                              
https://www.metacritic.com/game/3ds/10-in-1-arc...  10-in-1: Arcade Collection   
https://www.metacritic.com/game/3ds/1001-spikes                    1001 Spikes   
https://www.metacritic.com/game/3ds/101-dinopet...             101 DinoPets 3D   
https://www.metacritic.com/game/3ds/101-penguin...         101 Penguin Pets 3D   
https://www.metacritic.com/game/3ds/101-pony-pe...            101 Pony Pets 3D   

                                                    suggestions_count  \
url                                                                     
https://www.metacritic.com/game/3ds/10-in-1-arc...                109   
https://www.metacritic.com/game/3ds/1001-spikes                   370   
https://www.metacritic.com/game/3ds/101-dinopet...                153   
https://www.metacritic.com/game/3ds/101-penguin...                120   
https://www.metacritic.com/game/3ds/101-pony-pe...                149   

                                                        platform  \
url                                                                
https://www.metacritic.com/game/3ds/10-in-1-arc...  Nintendo 3DS   
https://www.metacritic.com/game/3ds/1001-spikes     Nintendo 3DS   
https://www.metacritic.com/game/3ds/101-dinopet...  Nintendo 3DS   
https://www.metacritic.com/game/3ds/101-penguin...  Nintendo 3DS   
https://www.metacritic.com/game/3ds/101-pony-pe...  Nintendo 3DS   

                                                                        genres  \
url                                                                              
https://www.metacritic.com/game/3ds/10-in-1-arc...                         NaN   
https://www.metacritic.com/game/3ds/1001-spikes     [Action, Adventure, Indie]   
https://www.metacritic.com/game/3ds/101-dinopet...                         NaN   
https://www.metacritic.com/game/3ds/101-penguin...                         NaN   
https://www.metacritic.com/game/3ds/101-pony-pe...                         NaN   

                                                                                               stores  \
url                                                                                                     
https://www.metacritic.com/game/3ds/10-in-1-arc...                                     Nintendo Store   
https://www.metacritic.com/game/3ds/1001-spikes     [Nintendo Store, Steam, PlayStation Store, Xbo...   
https://www.metacritic.com/game/3ds/101-dinopet...                                     Nintendo Store   
https://www.metacritic.com/game/3ds/101-penguin...                                     Nintendo Store   
https://www.metacritic.com/game/3ds/101-pony-pe...                                     Nintendo Store   

                                                                                                 tags  \
url                                                                                                     
https://www.metacritic.com/game/3ds/10-in-1-arc...  [Music, battle, fun, challenge, collection, Mi...   
https://www.metacritic.com/game/3ds/1001-spikes     [Singleplayer, Steam Achievements, Multiplayer...   
https://www.metacritic.com/game/3ds/101-dinopet...                                        Virtual Pet   
https://www.metacritic.com/game/3ds/101-penguin...                                        Virtual Pet   
https://www.metacritic.com/game/3ds/101-pony-pe...      [fun, Minigames, love, dance, fly, food, pet]   

                                                    esrb  critic_score  \
url                                                                      
https://www.metacritic.com/game/3ds/10-in-1-arc...   NaN           NaN   
https://www.metacritic.com/game/3ds/1001-spikes     Teen          73.0   
https://www.metacritic.com/game/3ds/101-dinopet...   NaN           NaN   
https://www.metacritic.com/game/3ds/101-penguin...   NaN           NaN   
https://www.metacri

We no longer need `url`, so we drop it using `reset_index` as we converted it to our index.

In [10]:
rawg_meta = rawg_meta.reset_index(drop = True)
rawg_meta.head()

name  suggestions_count      platform  \
0  10-in-1: Arcade Collection                109  Nintendo 3DS   
1                 1001 Spikes                370  Nintendo 3DS   
2             101 DinoPets 3D                153  Nintendo 3DS   
3         101 Penguin Pets 3D                120  Nintendo 3DS   
4            101 Pony Pets 3D                149  Nintendo 3DS   

                       genres  \
0                         NaN   
1  [Action, Adventure, Indie]   
2                         NaN   
3                         NaN   
4                         NaN   

                                              stores  \
0                                     Nintendo Store   
1  [Nintendo Store, Steam, PlayStation Store, Xbo...   
2                                     Nintendo Store   
3                                     Nintendo Store   
4                                     Nintendo Store   

                                                tags  esrb  critic_score  \
0  [Music, battle, fun, challenge, collection, Mi...   NaN           NaN   
1  [Singleplayer, Steam Achievements, Multiplayer...  Teen          73.0   
2                                        Virtual Pet   NaN           NaN   
3                                        Virtual Pet   NaN           NaN   
4      [fun, Minigames, love, dance, fly, food, pet]   NaN           NaN   

   critic_reviews  critic_pos  critic_mix  critic_neg user_score  \
0             NaN         NaN         NaN         NaN        3.3   
1             4.0         3.0         1.0         0.0        6.6   
2             NaN         NaN         NaN         NaN        tbd   
3             NaN         NaN         NaN         NaN        tbd   
4             NaN         NaN         NaN         NaN         10   

   user_reviews user_pos user_mix user_neg  
0           4.0      0.0      0.0      4.0  
1          15.0        9        2        4  
2           NaN      0.0      1.0      2.0  
3           NaN      0.0      1.0      0.0  
4           4.0      4.0      0.0      0.0

---
# 3. Merging RAWG and VGC Datasets
## 3.1. Platforms in `rawg_meta` VS `vgc_data` Datasets

In order to merge both datasets, the columns giving the information about the platform have to have the same format. We already know the column from `rawg_meta`. Let's see what `vgc_data` has:

In [11]:
vgc_data['Platform'].unique()

array(['All', 'Wii', 'NES', 'PC', 'NS', 'GB', 'DS', 'X360', 'SNES', 'PS3',
       'PS4', '3DS', 'PS2', 'GBA', 'GEN', 'N64', 'PS', 'XOne', 'WiiU',
       'XB', 'PSP', '2600', 'GC', 'GBC', 'PSN', 'PSV', 'DC', 'XBL', 'SAT',
       'SCD', '3DO', 'iOS', 'WS', 'Amig', 'VC', 'NG', 'WW', 'PCE', 'GG',
       'OSX', 'PCFX', 'Mob', 'And', 'Ouya', 'DSiW', 'MS', 'DSi', 'VB',
       'Linux', 'MSD', 'C128', 'AST', 'Lynx', '7800', '5200', 'S32X',
       'MSX', 'FMT', 'ACPC', 'C64', 'BRW', 'AJ', 'ZXS', 'NGage', 'GIZ',
       'WinP', 'iQue', 'Arc', 'ApII', 'Aco', 'BBCM', 'TG16', 'CDi',
       'CD32', 'XS', 'PS5', 'Int', 'CV'], dtype=object)

---

We can see that it doesn't match the `rawg_meta` format at all, which is:

In [12]:
rawg_meta['platform'].unique()

array(['Nintendo 3DS', 'Dreamcast', 'Nintendo DS', 'Nintendo DSi',
       'GameCube', 'Game Boy Advance', 'Nintendo 64', 'PC',
       'PlayStation 4', 'PlayStation 5', 'PlayStation', 'PlayStation 2',
       'PlayStation 3', 'PSP', 'Nintendo Switch', 'PS Vita', 'Wii U',
       'Wii', 'Xbox 360', 'Xbox One', 'Xbox Series S/X', 'Xbox'],
      dtype=object)

---
Let's first drop the undesired platforms (iOS, Android and Browser, as they are usually free and we don't want to study them in the same category as console games).

After that, we will filter out those games with sales < 0 and check if the unique platforms list gets reduced or not.

### 3.2. Dropping rows containing undesired platforms (free platforms)

In [13]:
# Check shape before dropping
vgc_data.shape[0]

59052

In [14]:
ios = vgc_data.index[vgc_data['Platform']=='iOS'].tolist()
android = vgc_data.index[vgc_data['Platform']=='And'].tolist()
browser = vgc_data.index[vgc_data['Platform']=='BRW'].tolist()
undesired = ios + android + browser

vgc_data = vgc_data.drop(index = undesired)\
                   .reset_index(drop = True)

In [15]:
# Check shape after dropping
vgc_data.shape[0]

57839

We reduced the dimension of the dataframe by a huge amount (~36% of the original).

In [16]:
vgc_data['Platform'].unique()

array(['All', 'Wii', 'NES', 'PC', 'NS', 'GB', 'DS', 'X360', 'SNES', 'PS3',
       'PS4', '3DS', 'PS2', 'GBA', 'GEN', 'N64', 'PS', 'XOne', 'WiiU',
       'XB', 'PSP', '2600', 'GC', 'GBC', 'PSN', 'PSV', 'DC', 'XBL', 'SAT',
       'SCD', '3DO', 'WS', 'Amig', 'VC', 'NG', 'WW', 'PCE', 'GG', 'OSX',
       'PCFX', 'Mob', 'Ouya', 'DSiW', 'MS', 'DSi', 'VB', 'Linux', 'MSD',
       'C128', 'AST', 'Lynx', '7800', '5200', 'S32X', 'MSX', 'FMT',
       'ACPC', 'C64', 'AJ', 'ZXS', 'NGage', 'GIZ', 'WinP', 'iQue', 'Arc',
       'ApII', 'Aco', 'BBCM', 'TG16', 'CDi', 'CD32', 'XS', 'PS5', 'Int',
       'CV'], dtype=object)

We can see that the number of platforms has also been reduced.

## 3.3. Checking Consoles

Let's discover now which console is which in this dataframe:

- **All**: All consoles available for the title.
- **Wii**: Nintendo Wii
- **NES**: Nintendo Entertainment System
- **PC**: PC
- **NS**: Nintendo Switch
- **GB**: Nintendo GameBoy
- **DS**: Nintendo DS
- **X360**: Xbox 360
- **SNES**: Super Nintendo Entertainment System
- **PS3**: PlayStation 3
- **PS4**: PlayStation 4
- **3DS**: Nintendo 3DS
- **PS2**: PlayStation 2
- **GBA**: Nintendo GameBoy Advance
- **GEN**: Sega Genesis
- **N64**: Nintendo 64
- **PS**: PlayStation
- **XOne**: Xbox One
- **WiiU**: Nintendo Wii U
- **XB**: Xbox
- **PSP**: PlayStation Portable
- **2600**: Atari 2600
- **GC**: Nintendo GameCube
- **GBC**: Nintendo GameBoy Color
- **PSN**: PlayStation Network
- **PSV**: PlayStation Vita
- **DC**: Dreamcast
- **XBL**: Xbox Live
- **SAT**: Sega Saturn
- **SCD**: Sega CD System
- **3DO**: 3DO Interactive Multiplayer
- **WS**: Wonder Swan
- **Amig**: Commodore Amiga
- **VC**: Nintendo Virtual Console
- **NG**: Neo Geo
- **WW**: Nintendo WiiWare (online shop) --> **Count as Wii**
- **PCE**: PC Engine
- **GG**: Game Gear
- **OSX**: Apple OSX --> **Count as PC**
- **PCFX**: PC-FX
- **Mob**: Mini Mobile

This list, brought into a table would look like this:

| Multiplatform | PC | Sony | MicroSoft | Nintendo | Others |
|---------------|----|------|-----------|----------|--------|
| All | PC<br>OSX | PlayStation 4<br>PlayStation 3<br>PlayStation 2<br>PlayStation<br>PlayStation Network<br>PlayStation Portable<br>PlayStation Vita | Xbox One<br>Xbox 360<br>Xbox<br>Xbox Live | Switch<br>Wii U<br>Wii<br>GameCube<br>Nintendo 64<br>SNES<br>NES<br>Virtual Console<br>3DS<br>DS<br>GBA<br>GBC<br>GB | Sega Genesis<br>Sega Saturn<br>Sega CD System<br>Atari 2600<br>Dreamcast<br>3DO Interactive Multiplayer<br>Wonder Swan<br>Commodore Amiga<br>Neo Geo<br>PC Engine<br>Game Gear<br>PC-FX<br>Mini Mobile |


## 3.4. Transforming Consoles

In order to merge both datasets, we should create a new column called `join_platform` and assigning values on the following dictionaries using each dataset's platforms column as keys:

In [17]:
# VGC

platforms_vgc = {
    
    # All
    
    "All": "All",
    
    # PC
    
    "PC": "PC",
    "OSX": "PC",
    
    # Sony

    "PS4": "PS4",
    "PS3": "PS3",
    "PS2": "PS2",
    "PS": "PS",
    "PSN": "PSN",
    "PSV": "Vita",
    "PSP": "PSP",
    
    # Microsoft
    
    "XOne": "XOne",
    "X360": "X360",
    "XB": "Xbox",
    "XBL": "Xbox Live",
    
    # Nintendo
    
    "NS": "Switch",
    "WiiU": "Wii U",
    "Wii": "Wii",
    "WW": "Wii",
    "VC": "VC",
    "GC": "GCube",
    "N64": "N64",
    "SNES": "SNES",
    "NES": "NES",
    "3DS": "3DS",
    "DS": "DS",
    "GBA": "GBA",
    "GBC": "GBC",
    "GB": "GB",
    
    # Others
    
    "DC": "Dreamcast",
    "GEN": "Genesis",
    "2600": "Atari 2600",
    "SAT": "Saturn",
    "SCD": "Sega CD",
    "3DO": "3DO",
    "WS": "WSwan",
    "Amig": "Amiga",
    "NG": "NeoGeo",
    "PCE": "PC Engine",
    "GG": "Game Gear",
    "PCFX": "PCFX",
    "Mob": "Mini Mobile"
    
}

In [18]:
#RAWG_Meta

platforms_rawg = {
    
    # Sony
    
    "PlayStation 5": "PS5",
    "PlayStation 4": "PS4",
    "PlayStation 3": "PS3",
    "PlayStation 2": "PS2",
    "PlayStation": "PS",
    "PS Vita": "Vita",
    "PSP": "PSP",
    
    # Microsoft
    
    "Xbox One": "XOne",
    "Xbox Series S/X": "XS/X",
    "Xbox 360": "X360",
    "Xbox": "Xbox",
    
    # Nintendo
    
    "Nintendo Switch": "Switch",
    "Wii U": "Wii U",
    "Wii": "Wii",
    "GameCube": "GCube",
    "Nintendo 64": "N64",
    "Nintendo 3DS": "3DS",
    "Nintendo DS": "DS",
    "Nintendo DSi": "DS",
    "Game Boy Advance": "GBA",
    
    # Others
    
    "PC": "PC",
    "Dreamcast": "Dreamcast"
    
}

In [19]:
vgc_data['Platform'] = vgc_data['Platform'].map(platforms_vgc)
rawg_meta['platform'] = rawg_meta['platform'].map(platforms_rawg)

vgc_data

Title Platform  \
0                                    Minecraft      All   
1                           Grand Theft Auto V      All   
2                                   Wii Sports      Wii   
3                PlayerUnknown's Battlegrounds      All   
4                            Super Mario Bros.      NES   
...                                        ...      ...   
57834       Zombieland: Double Tap - Road Trip       PC   
57835                                Zombillie   Switch   
57836  Zone of the Enders: The 2nd Runner MARS       PC   
57837   Zoo Tycoon: Ultimate Animal Collection     XOne   
57838   Zoo Tycoon: Ultimate Animal Collection       PC   

                        Publisher                   Developer    Release  \
0                          Mojang                   Mojang AB 2011-11-18   
1                  Rockstar Games              Rockstar North 2013-09-17   
2                        Nintendo                Nintendo EAD 2006-11-19   
3                PUBG Corporation            PUBG Corporation 2017-12-17   
4                        Nintendo                Nintendo EAD 1985-10-18   
...                           ...                         ...        ...   
57834      GameMill Entertainment       High Voltage Software 2019-10-15   
57835  Forever Entertainment S.A.  Forever Entertainment S.A. 2018-03-29   
57836                      Konami                     Cygames 2018-06-09   
57837           Microsoft Studios       Frontier Developments 2017-10-31   
57838                  THQ Nordic       Frontier Developments 2018-09-14   

        Sales  
0      200.00  
1      140.00  
2       82.90  
3       70.00  
4       40.24  
...       ...  
57834    0.17  
57835    0.17  
57836    0.17  
57837    0.17  
57838    0.17  

[57839 rows x 6 columns]

In [20]:
rawg_meta

name  suggestions_count platform  \
0         10-in-1: Arcade Collection                109      3DS   
1                        1001 Spikes                370      3DS   
2                    101 DinoPets 3D                153      3DS   
3                101 Penguin Pets 3D                120      3DS   
4                   101 Pony Pets 3D                149      3DS   
...                              ...                ...      ...   
93476                        Xyanide                125     Xbox   
93477                          Yager                114     Xbox   
93478  Yu-Gi-Oh! The Dawn of Destiny                160     Xbox   
93479     Zapper: One Wicked Cricket                333     Xbox   
93480                        Zathura                227     Xbox   

                           genres  \
0                             NaN   
1      [Action, Adventure, Indie]   
2                             NaN   
3                             NaN   
4                             NaN   
...                           ...   
93476                      Action   
93477        [Action, Simulation]   
93478                    Strategy   
93479                         NaN   
93480                   Adventure   

                                                  stores  \
0                                         Nintendo Store   
1      [Nintendo Store, Steam, PlayStation Store, Xbo...   
2                                         Nintendo Store   
3                                         Nintendo Store   
4                                         Nintendo Store   
...                                                  ...   
93476                                                NaN   
93477                                                NaN   
93478                                                NaN   
93479                                                NaN   
93480                                                NaN   

                                                    tags          esrb  \
0      [Music, battle, fun, challenge, collection, Mi...           NaN   
1      [Singleplayer, Steam Achievements, Multiplayer...          Teen   
2                                            Virtual Pet           NaN   
3                                            Virtual Pet           NaN   
4          [fun, Minigames, love, dance, fly, food, pet]           NaN   
...                                                  ...           ...   
93476                        [exclusive, true exclusive]           NaN   
93477                                                NaN           NaN   
93478                        [exclusive, true exclusive]           NaN   
93479                                                NaN           NaN   
93480                                                NaN  Everyone 10+   

       critic_score  critic_reviews  critic_pos  critic_mix  critic_neg  \
0               NaN             NaN         NaN         NaN         NaN   
1              73.0             4.0         3.0         1.0         0.0   
2               NaN             NaN         NaN         NaN         NaN   
3               NaN             NaN         NaN         NaN         NaN   
4               NaN             NaN         NaN         NaN         NaN   
...             ...             ...         ...         ...         ...   
93476          70.0            16.0         6.0         9.0         1.0   
93477          70.0            19.0         7.0        12.0         0.0   
93478           NaN             NaN         NaN         NaN         NaN   
93479           NaN             NaN         NaN         NaN         NaN   
93480          48.0            15.0         0.0         7.0         8.0   

      user_score  user_reviews user_pos user_mix user_neg  
0            3.3           4.0      0.0      0.0      4.0  
1            6.6          15.0        9        2        4  
2            tbd           NaN      0.0      1.0      2.0  
3            tbd   

---
## 3.5. Merging

We want to join both dataframes taking these into account:

- Game title
- Platform

But if we did a regular join, this would be the result:

In [21]:
vgc_data.set_index('Title').join(rawg_meta.set_index('name'))

Platform                Publisher  \
"Nuke It"                            PC            CrystalVision   
#IDARB                             XOne  Other Ocean Interactive   
#killallzombies                     PS4              Beatshapers   
'70s Robot Anime: Geppy-X            PS                    Aroma   
'98 Koshien                          PS          Magical Company   
...                                 ...                      ...   
yOm                           Xbox Live                Microsoft   
yOm_fury                      Xbox Live                Microsoft   
¡Shin Chan Flipa en colores!         DS                505 Games   
じんるいのみなさまへ                       Switch                  Unknown   
レミロア~少女と異世界と魔導書                  Switch                  Unknown   

                                            Developer    Release  Sales  \
"Nuke It"                               CrystalVision 1998-01-01   0.17   
#IDARB                        Other Ocean Interactive 2015-01-02   0.17   
#killallzombies                           Beatshapers 2015-01-01   0.17   
'70s Robot Anime: Geppy-X                       Aroma 1999-05-27   0.17   
'98 Koshien                           Magical Company 1998-06-18   0.41   
...                                               ...        ...    ...   
yOm                                            jojito 2009-10-21   0.17   
yOm_fury                                       jojito 2009-12-11   0.17   
¡Shin Chan Flipa en colores!             Inti Creates 2007-11-16   0.14   
じんるいのみなさまへ                       Nippon Ichi Software        NaT   0.17   
レミロア~少女と異世界と魔導書                                 Pikii        NaT   0.17   

                              suggestions_count platform genres      stores  \
"Nuke It"                                   NaN      NaN    NaN         NaN   
#IDARB                                    182.0     XOne    NaN  Xbox Store   
#killallzombies                             NaN      NaN    NaN         NaN   
'70s Robot Anime: Geppy-X                   NaN      NaN    NaN         NaN   
'98 Koshien                                 NaN      NaN    NaN         NaN   
...                                         ...      ...    ...         ...   
yOm                                         NaN      NaN    NaN         NaN   
yOm_fury                                    NaN      NaN    NaN         NaN   
¡Shin Chan Flipa en colores!                NaN      NaN    NaN         NaN   
じんるいのみなさまへ                                  NaN      NaN    NaN         NaN   
レミロア~少女と異世界と魔導書                             NaN      NaN    NaN         NaN   

                                                                           tags  \
"Nuke It"                                                                   NaN   
#IDARB                        [Music, Story, battle, future, strange, ball, ...   
#killallzombies                                                             NaN   
'70s Robot Anime: Geppy-X                                                   NaN   
'98 Koshien                                                                 NaN   
...                                                                         ...   
yOm                                                                         NaN   
yOm_fury                                                                    NaN   
¡Shin Chan Flipa en colores!                                                NaN   
じんるいのみなさまへ                                                                  NaN   
レミロア~少女と異世界と魔導書                                                             NaN   

                              ... critic_score  critic_reviews  critic_pos  \
"Nuke It"                     ...          NaN             NaN         NaN   
#IDARB                        ...         77.0            31.0        23.0   
#killallzombies               ...          NaN             NaN         NaN   
'70s Robot Anime: Geppy-X     ...          NaN         

We can see that `Platform` and `platform` do not match each other (and thus, now the dataset is bigger than it should be).

To solve this problem, we will lowercase the game title and add the platform string to it:

In [22]:
rawg_meta['join_col'] = rawg_meta['name'] + ' ' + rawg_meta['platform']
rawg_meta = rawg_meta.set_index('join_col')

In [23]:
vgc_data['join_col'] = vgc_data['Title'] + ' ' + vgc_data['Platform']
vgc_data = vgc_data.set_index('join_col')

We drop the `join_col` indexes and the `name` and `platform` columns as we join both dataframes.

In [24]:
data = vgc_data.join(rawg_meta).reset_index(drop = True).drop(['name','platform'], axis = 1)
data

Title Platform                  Publisher  \
0                      "Nuke It"       PC              CrystalVision   
1                         #IDARB     XOne    Other Ocean Interactive   
2                #killallzombies      PS4                Beatshapers   
3      '70s Robot Anime: Geppy-X       PS                      Aroma   
4                    '98 Koshien       PS            Magical Company   
...                          ...      ...                        ...   
57846        XCOM: Enemy Unknown      NaN          Feral Interactive   
57847      Yakuza: Like A Dragon      NaN                       Sega   
57848      Yakuza: Like A Dragon      NaN                       Sega   
57849                     Zaxxon      NaN                     Coleco   
57850          Zelda's Adventure      NaN  Philips Interactive Media   

                     Developer    Release  Sales  suggestions_count genres  \
0                CrystalVision 1998-01-01   0.17                NaN    NaN   
1      Other Ocean Interactive 2015-01-02   0.17              182.0    NaN   
2                  Beatshapers 2015-01-01   0.17                NaN    NaN   
3                        Aroma 1999-05-27   0.17                NaN    NaN   
4              Magical Company 1998-06-18   0.41                NaN    NaN   
...                        ...        ...    ...                ...    ...   
57846            Firaxis Games 2014-06-19   0.17                NaN    NaN   
57847                     Sega 2020-10-11   0.17                NaN    NaN   
57848                     Sega 2020-01-12   0.17                NaN    NaN   
57849                   Coleco 1983-01-01   0.17                NaN    NaN   
57850      Viridis Corporation 1994-05-06   0.17                NaN    NaN   

           stores                                               tags  ...  \
0             NaN                                                NaN  ...   
1      Xbox Store  [Music, Story, battle, future, strange, ball, ...  ...   
2             NaN                                                NaN  ...   
3             NaN                                                NaN  ...   
4             NaN                                                NaN  ...   
...           ...                                                ...  ...   
57846         NaN                                                NaN  ...   
57847         NaN                                                NaN  ...   
57848         NaN                                                NaN  ...   
57849         NaN                                                NaN  ...   
57850         NaN                                                NaN  ...   

      critic_score  critic_reviews  critic_pos  critic_mix  critic_neg  \
0              NaN             NaN         NaN         NaN         NaN   
1             77.0            31.0        23.0         8.0         0.0   
2              NaN             NaN         NaN         NaN         NaN   
3              NaN             NaN         NaN         NaN         NaN   
4              NaN             NaN         NaN         NaN         NaN   
...            ...             ...         ...         ...         ...   
57846          NaN             NaN         NaN         NaN         NaN   
57847          NaN             NaN         NaN         NaN         NaN   
57848          NaN             NaN         NaN         NaN         NaN   
57849          NaN             NaN         NaN         NaN         NaN   
57850          NaN             NaN         NaN         NaN         NaN   

       user_score user_reviews  user_pos user_mix user_neg  
0             NaN          NaN       NaN      NaN      NaN  
1             6.9         88.0        47       26       15  
2             NaN          NaN       NaN      NaN      NaN  
3             NaN          NaN       NaN      NaN      NaN  
4             NaN          NaN       NaN      NaN      NaN  
...           ...          ...       ...     

## 3.6. Renaming Columns
We rename the columns from the `rawg_meta` dataframe to match the column names from the VGC dataframe.

In [25]:
rename = {
    "suggestions_count": "Suggest_count",
    "genres": "Genres",
    "stores": "Stores",
    "tags": "Tags",
    "esrb": "ESRB",
    "critic_score": "C_Score",
    "critic_reviews": "C_Reviews",
    "critic_pos": "C_Positive",
    "critic_mix": "C_Mixed",
    "critic_neg": "C_Negative",
    "user_score": "U_Score",
    "user_reviews": "U_Reviews",
    "user_pos": "U_Positive",
    "user_mix": "U_Mixed",
    "user_neg": "U_Negative"
}

In [26]:
data = data.rename(columns = rename)
data.head()

Title Platform                Publisher  \
0                  "Nuke It"       PC            CrystalVision   
1                     #IDARB     XOne  Other Ocean Interactive   
2            #killallzombies      PS4              Beatshapers   
3  '70s Robot Anime: Geppy-X       PS                    Aroma   
4                '98 Koshien       PS          Magical Company   

                 Developer    Release  Sales  Suggest_count Genres  \
0            CrystalVision 1998-01-01   0.17            NaN    NaN   
1  Other Ocean Interactive 2015-01-02   0.17          182.0    NaN   
2              Beatshapers 2015-01-01   0.17            NaN    NaN   
3                    Aroma 1999-05-27   0.17            NaN    NaN   
4          Magical Company 1998-06-18   0.41            NaN    NaN   

       Stores                                               Tags  ... C_Score  \
0         NaN                                                NaN  ...     NaN   
1  Xbox Store  [Music, Story, battle, future, strange, ball, ...  ...    77.0   
2         NaN                                                NaN  ...     NaN   
3         NaN                                                NaN  ...     NaN   
4         NaN                                                NaN  ...     NaN   

   C_Reviews  C_Positive  C_Mixed  C_Negative  U_Score U_Reviews  U_Positive  \
0        NaN         NaN      NaN         NaN      NaN       NaN         NaN   
1       31.0        23.0      8.0         0.0      6.9      88.0          47   
2        NaN         NaN      NaN         NaN      NaN       NaN         NaN   
3        NaN         NaN      NaN         NaN      NaN       NaN         NaN   
4        NaN         NaN      NaN         NaN      NaN       NaN         NaN   

  U_Mixed U_Negative  
0     NaN        NaN  
1      26         15  
2     NaN        NaN  
3     NaN        NaN  
4     NaN        NaN  

[5 rows x 21 columns]

# 4. Checking Values in Critic and User-Related Columns 

In [27]:
check_c = data.filter(like = 'C_', axis = 1)
check_u = data.filter(like = 'U_', axis = 1)

In [28]:
check_c.dtypes

C_Score       float64
C_Reviews     float64
C_Positive    float64
C_Mixed       float64
C_Negative    float64
dtype: object

In [29]:
check_u.dtypes

U_Score        object
U_Reviews     float64
U_Positive     object
U_Mixed        object
U_Negative     object
dtype: object

We see that the **User-related columns have strings in them**. We must transform them to numeric.

- `U_Score` seems to have some of their values as `tbd`, so we will transform them to NaN.
- The rest of user columns contain commas, so we are going to replace them with a blank space.

Then we are filtering the dataframe to apply a `pd.to_numeric` transformation to all of them.

In [30]:
# We replace 'tbd' with NaN
data['U_Score'] = data['U_Score'].map(lambda x: x if x != 'tbd' else np.nan)

# We remove commas from thousands
data['U_Positive'] = data['U_Positive'].str.replace(',','')
data['U_Mixed'] = data['U_Mixed'].str.replace(',','')
data['U_Negative'] = data['U_Negative'].str.replace(',','')

In [31]:
data[data.filter(like = 'U_', axis = 1).columns] = data.filter(like = 'U_', axis = 1).apply(pd.to_numeric)

data['U_Score'] = data['U_Score'] * 10 # We multiply this column by 10 so that it is in the same scale as C_Score (out of 100 instead of 10).

We check if the transformation was done correctly.

In [32]:
data.filter(like = 'U_', axis = 1).dtypes

U_Score       float64
U_Reviews     float64
U_Positive    float64
U_Mixed       float64
U_Negative    float64
dtype: object

# 5. Saving Dataset

In [33]:
data.to_csv("Files/04_merged_data.csv", encoding='utf-8', index=False)

---
# 6. Dealing with some features

Some columns with categorical values such as:

- `Publisher`

- `Developer`

- `Platform`

- `Genres`

- `Stores`

- `Tags`

are potentially a problem, as applying One-Hot Encoding to each of them would cause **problems in dimensionality**, mainly due to `Tags` containing **more than 1000 unique values**.

## 6.1. Counting Unique Values

Let's quicky check how many unique values they contain:

In [34]:
def unique_nested_values(a_col, a_list):
    
    '''
    Given a pd.Series object and a list, retuns the count of unique values inside the Series object in the input list.
    
    a_col: Column of interest.
    a_list: List where we want the counts to be stored on.
    '''
    
    col = a_col
    result = a_list
    
    if type(col) != list:
        if col not in result:
            result.append(col)
    else:
        for element in col:
            if element not in result:
                result.append(element)
    
    return result

In [35]:
genres = []
stores = []
tags = []

for idx, row in data.iterrows():
    
    genre = row['Genres']
    store = row['Stores']
    tag = row['Tags']

    unique_nested_values(genre, genres)
    unique_nested_values(store, stores)
    unique_nested_values(tag, tags)

In [36]:
print(str(data['Platform'].value_counts().shape[0]) + ' unique values')
data['Platform'].value_counts()

39 unique values


PC             12314
PS2             3566
DS              3297
PS              2706
PS4             2255
Wii             2140
Xbox Live       2116
PSN             2004
PS3             1874
PSP             1807
Switch          1758
X360            1704
GBA             1657
GB              1600
XOne            1489
SNES            1210
3DS             1186
NES             1107
Vita            1083
Xbox             984
Genesis          805
Saturn           739
GCube            663
Dreamcast        655
VC               654
All              637
Wii U            545
Atari 2600       501
PC Engine        459
N64              393
Game Gear        334
3DO              304
Sega CD          218
WSwan            199
NeoGeo           172
PCFX              67
Amiga             30
Mini Mobile       19
GBC               15
Name: Platform, dtype: int64

In [37]:
print(str(data['Publisher'].value_counts().shape[0]) + ' unique values')
data['Publisher'].value_counts()

3053 unique values


Unknown               5710
Sega                  2113
Ubisoft               1579
Activision            1524
Electronic Arts       1520
                      ... 
2DArray                  1
Platine Dispositif       1
Techiku                  1
Haydee Interactive       1
tri-Ace                  1
Name: Publisher, Length: 3053, dtype: int64

---

In [38]:
print(str(data['Developer'].value_counts().shape[0]) + ' unique values')
data['Developer'].value_counts()

8371 unique values


Unknown                  3604
Konami                    930
Sega                      857
Capcom                    758
Namco                     427
                         ... 
MuckyBaby Productions       1
Black Tower Studios         1
PheroseSoft                 1
fyto                        1
AftercastGames              1
Name: Developer, Length: 8371, dtype: int64

---

In [39]:
print(str(len(genres)) + ' unique values')
pd.Series(genres)

20 unique values


0                       NaN
1                       RPG
2                    Action
3                   Shooter
4                    Racing
5                     Indie
6                    Arcade
7                  Strategy
8               Board Games
9     Massively Multiplayer
10                   Puzzle
11                Adventure
12                   Sports
13                   Casual
14                   Family
15               Simulation
16               Platformer
17              Educational
18                 Fighting
19                     Card
dtype: object

---

In [40]:
print(str(len(stores)) + ' unique values')
pd.Series(stores)

11 unique values


0                   NaN
1            Xbox Store
2     PlayStation Store
3                 Steam
4        Xbox 360 Store
5        Nintendo Store
6             App Store
7               itch.io
8                   GOG
9           Google Play
10           Epic Games
dtype: object

---

In [41]:
print(str(len(tags)) + ' unique values')
pd.Series(tags)

1376 unique values


0                          NaN
1                        Music
2                        Story
3                       battle
4                       future
                 ...          
1371              Card Battler
1372    two-player multiplayer
1373                       zoo
1374                     rally
1375             Game Creation
Length: 1376, dtype: object

---

So we have:

|**Column** |**Unique Values**|**Comments**                                  |
|-----------|-----------------|----------------------------------------------|
|*Platform* |39               |                                              |
|*Publisher*|3053             |Will cause dimensionality problem!            |
|*Developer*|8371             |Will cause dimensionality problem!            |
|*Genres*   |20               |                                              |
|*Stores*   |11               |We should remove *App Store* and *Google Play*|
|*Tags*     |1379             |Will cause dimensionality problem!            |


As we can see, having a dataframe with >1500 columns does not look like a good idea. Thus, we have to take some decisions regarding these columns!

## 6.2. What to do with categorical columns

- What do we do with the `Platform` column?
    
    - Remove `All`s, as they take every platform for the title available at the moment (and sometimes, the sum does not match the figures in here).
    
    - Convert the platforms into the company selling the platform (e.g. 'Playstation 4' -> 'Sony', 'Wii' -> 'Nintendo'), summing the sales for the same company platform.
    
    
- What do we do with the `Genres` column?

    - As there are not many unique values, One-Hot Encoding is an option.
    
 
- What do we do with the `Stores` column?

    - Same as the `Genres` column.


- What do we do with the `Tags` column?

    - We have to research what we can do in this case.
    
    - One option found is using the top-10 most frequent tags for One-Hot Encoding.
    
    - Not using tags for the analysis.


- What do we do with the `Publisher` column?

    - Same thing as `Tags` column.
    
    - Target encoding the top 30~50 values.


- What do we do with the `Developer` column?

    - Same thing as `Tags` column.
    
    - Target encoding the top 30~50 values.